src/HybridSolver.cpp

```diff
@@ -20,673 +20,275 @@
 */
 
 #include "HybridSolver.hpp"
-#include "utilities.hpp"
-#include "MatrixUtilities.hpp"
-
-using std::unique_ptr;
 
 namespace smoothg
 {
 
-void HybridSolver::BuildFineLevelLocalMassMatrix(
-    const mfem::SparseMatrix& vertex_edge,
-    const mfem::SparseMatrix& M,
-    std::vector<mfem::Vector>& M_el)
-{
-    const int nvertices = vertex_edge.Height();
-    M_el.resize(nvertices);
-
-    mfem::SparseMatrix edge_vertex(smoothg::Transpose(vertex_edge));
-    mfem::Array<int> local_edgedof;
-
-    const mfem::Vector M_data(M.GetData(), M.Height());
-    for (int i = 0; i < nvertices; i++)
-    {
-        GetTableRow(vertex_edge, i, local_edgedof);
-        const int nlocal_edgedof = local_edgedof.Size();
-
-        mfem::Vector& Mloc(M_el[i]);
-        Mloc.SetSize(nlocal_edgedof);
-
-        for (int j = 0; j < nlocal_edgedof; j++)
-        {
-            const int edgedof = local_edgedof[j];
-
-            if (edge_vertex.RowSize(edgedof) == 2)
-                Mloc(j) = M_data[edgedof] / 2;
-            else
-                Mloc(j) = M_data[edgedof];
-        }
-    }
-}
-
-HybridSolver::HybridSolver(MPI_Comm comm,
-                           const MixedMatrix& mgL,
-                           const mfem::SparseMatrix* face_bdrattr,
-                           const mfem::Array<int>* ess_edge_dofs,
-                           bool spectralAMGe)
+HybridSolver::HybridSolver(const MixedMatrix& mgl)
     :
-    MixedLaplacianSolver(mgL.get_blockoffsets()),
-    comm_(comm),
-    D_(mgL.getD()),
-    W_(mgL.getW()),
-    spectralAMGe_(spectralAMGe),
-    use_w_(mgL.CheckW())
+    MGLSolver(mgl),
+    agg_vertexdof_(mgl.GetAggVertexDof()),
+    agg_edgedof_(mgl.GetElemDof()),
+    num_aggs_(agg_edgedof_.Rows()),
+    num_edge_dofs_(agg_edgedof_.Cols()),
+    num_multiplier_dofs_(mgl.GetFaceFaceDof().Cols()),
+    MinvDT_(num_aggs_), MinvCT_(num_aggs_),
+    AinvDMinvCT_(num_aggs_), Ainv_(num_aggs_),
+    hybrid_elem_(num_aggs_), Ainv_f_(num_aggs_),
+    agg_weights_(num_aggs_, 1.0)
 {
-    MPI_Comm_rank(comm, &myid_);
+    SparseMatrix edgedof_multiplier = MakeEdgeDofMultiplier();
+    SparseMatrix multiplier_edgedof = edgedof_multiplier.Transpose();
+    const std::vector<int>& j_multiplier_edgedof = multiplier_edgedof.GetIndices();
 
-    const int nvertices = D_.Height();
+    agg_multiplier_ = agg_edgedof_.Mult(edgedof_multiplier);
 
-    // TODO(gelever1): use operator= when mfem version is updated
-    mfem::SparseMatrix tmp = SparseIdentity(nvertices);
-    Agg_vertexdof_.Swap(tmp);
+    ParMatrix edge_td_d = mgl.EdgeTrueEdge().Transpose();
+    ParMatrix edge_edge = mgl.EdgeTrueEdge().Mult(edge_td_d);
+    ParMatrix edgedof_multiplier_d(comm_, std::move(edgedof_multiplier));
+    ParMatrix multiplier_d_td_d = parlinalgcpp::RAP(edge_edge, edgedof_multiplier_d);
 
-    Agg_edgedof_.MakeRef(D_);
-    const mfem::SparseMatrix edge_edgedof;
+    multiplier_d_td_ = MakeEntityTrueEntity(multiplier_d_td_d);
```

Comment:
Some suggestion (again probably another PR): line 46-51 here can group into one function? Also, seems computing `trueedgedof_multiplier` = `edge_td_d` * `edgedof_multiplier_d` and then multiply its transpose to it is a faster way to get `multiplier_d_td_d`? Again, the saving should be minimal.